tidb-server/server/conn.go

```diff
@@ -0,0 +1,469 @@
+	data = append(data, dumpLengthEncodedInt(uint64(cc.ctx.AffectedRows()))...)
+	data = append(data, dumpLengthEncodedInt(uint64(cc.ctx.LastInsertID()))...)
+	if cc.capability&mysql.ClientProtocol41 > 0 {
+		data = append(data, dumpUint16(cc.ctx.Status())...)
+		data = append(data, dumpUint16(cc.ctx.WarningCount())...)
+	}
+
+	err := cc.writePacket(data)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(cc.flush())
+}
+
+func (cc *clientConn) writeError(e error) error {
+	var m *mysql.SQLError
+	var ok bool
+	if m, ok = e.(*mysql.SQLError); !ok {
+		m = mysql.NewError(mysql.ErUnknownError, e.Error())
+	}
+
+	data := make([]byte, 4, 16+len(m.Message))
+	data = append(data, mysql.ErrHeader)
+	data = append(data, byte(m.Code), byte(m.Code>>8))
+	if cc.capability&mysql.ClientProtocol41 > 0 {
+		data = append(data, '#')
+		data = append(data, m.State...)
+	}
+
+	data = append(data, m.Message...)
+
+	err := cc.writePacket(data)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(cc.flush())
+}
+
+func (cc *clientConn) writeEOF() error {
+	data := cc.alloc.AllocWithLen(4, 9)
+
+	data = append(data, mysql.EOFHeader)
+	if cc.capability&mysql.ClientProtocol41 > 0 {
+		data = append(data, dumpUint16(cc.ctx.WarningCount())...)
+		data = append(data, dumpUint16(cc.ctx.Status())...)
+	}
+
+	err := cc.writePacket(data)
+	return errors.Trace(err)
+}
+
+func (cc *clientConn) handleQuery(sql string) (err error) {
+	rs, err := cc.ctx.Execute(sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if rs != nil {
+		err = cc.writeResultset(rs, false)
+	} else {
+		err = cc.writeOK()
+	}
+	return errors.Trace(err)
+}
+
+func (cc *clientConn) handleFieldList(sql string) (err error) {
+	parts := strings.Split(sql, "\x00")
+	columns, err := cc.ctx.FieldList(parts[0])
+	if err != nil {
+		return
+	}
+	data := make([]byte, 4, 1024)
+	for _, v := range columns {
+		data = data[0:4]
+		data = append(data, v.Dump(cc.alloc)...)
+		if err := cc.writePacket(data); err != nil {
+			return err
+		}
+	}
+	if err := cc.writeEOF(); err != nil {
+		return err
+	}
+	return errors.Trace(cc.flush())
+}
+
+func (cc *clientConn) writeResultset(rs ResultSet, binary bool) error {
+	defer rs.Close()
+	// We need to call Next before we get columns.
+	// Otherwise, we will get incorrect columns info.
+	row, err := rs.Next()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	columns, err := rs.Columns()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	columnLen := dumpLengthEncodedInt(uint64(len(columns)))
+	data := cc.alloc.AllocWithLen(4, 1024)
+	data = append(data, columnLen...)
+	if err := cc.writePacket(data); err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, v := range columns {
+		data = data[0:4]
+		data = append(data, v.Dump(cc.alloc)...)
+		if err := cc.writePacket(data); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	if err := cc.writeEOF(); err != nil {
+		return errors.Trace(err)
+	}
+
+	for {
+		if err != nil {
```

Comment:
because err if set at the end of the for loop, and I think err check and nil row check should be put together.